src/goto-instrument/contracts/assigns.cpp

```diff
@@ -9,144 +9,177 @@ Date: July 2021
+
+    INVARIANT(
+      size.has_value(),
+      "`sizeof` must always be computable on l-value assigns clause targets.");
+
+    return {typecast_exprt::conditional_cast(
+              address_of_exprt{expr}, pointer_type(char_type())),
+            typecast_exprt::conditional_cast(size.value(), signed_size_type())};
+  }
+
+  UNREACHABLE;
+}
+
+const symbolt assigns_clauset::conditional_address_ranget::generate_new_symbol(
+  const typet &type,
+  const source_locationt &location) const
+{
+  return new_tmp_symbol(
+    type,
+    location,
+    parent.symbol_table.lookup_ref(parent.function_name).mode,
+    parent.symbol_table);
+}
 
-assigns_clauset::targett::targett(
+assigns_clauset::conditional_address_ranget::conditional_address_ranget(
   const assigns_clauset &parent,
   const exprt &expr)
-  : address(address_of_exprt(normalize(expr))),
-    id(expr.id()),
-    parent(parent)
+  : source_expr(expr),
+    location(expr.source_location()),
+    parent(parent),
+    slice(normalize_to_slice(expr, parent.ns)),
+    validity_condition_var(
+      generate_new_symbol(bool_typet(), location).symbol_expr()),
+    lower_bound_address_var(
+      generate_new_symbol(slice.first.type(), location).symbol_expr()),
+    upper_bound_address_var(
+      generate_new_symbol(slice.first.type(), location).symbol_expr())
 {
 }
 
-exprt assigns_clauset::targett::normalize(const exprt &expr)
+goto_programt
+assigns_clauset::conditional_address_ranget::generate_snapshot_instructions()
+  const
 {
-  if(expr.id() != ID_address_of)
-    return expr;
+  goto_programt instructions;
+
+  instructions.add(goto_programt::make_decl(validity_condition_var, location));
+  instructions.add(goto_programt::make_decl(lower_bound_address_var, location));
+  instructions.add(goto_programt::make_decl(upper_bound_address_var, location));
 
-  const auto &object = to_address_of_expr(expr).object();
-  if(object.id() != ID_index)
-    return object;
+  goto_programt skip_program;
+  const auto skip_target = skip_program.add(goto_programt::make_skip(location));
 
-  return to_index_expr(object).array();
+  const auto validity_check_expr =
+    and_exprt{all_dereferences_are_valid(source_expr, parent.ns),
+              w_ok_exprt{slice.first, slice.second}};
+  instructions.add(goto_programt::make_assignment(
+    validity_condition_var, validity_check_expr, location));
+
+  instructions.add(goto_programt::make_goto(
+    skip_target, not_exprt{validity_condition_var}, location));
+
+  instructions.add(goto_programt::make_assignment(
+    lower_bound_address_var, slice.first, location));
+
+  instructions.add(goto_programt::make_assignment(
+    upper_bound_address_var,
+    minus_exprt{plus_exprt{slice.first, slice.second},
+                from_integer(1, slice.second.type())},
+    location));
+
+  instructions.destructive_append(skip_program);
+  return instructions;
+}
+
+const exprt
+assigns_clauset::conditional_address_ranget::generate_unsafe_inclusion_check(
```

Comment:
Why unsafe?

Reply:
Because validity is not checked for `lhs`.